test/api/transport/transport.handshake.js

```diff
@@ -0,0 +1,275 @@
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Missing required property');
+					done();
+				});
+			});
+
+			it('should fail with undefined headers', function (done) {
+				validClientSocketOptions.query = undefined;
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Missing required property');
+					done();
+				});
+			});
+
+			it('should fail with empty headers', function (done) {
+				validClientSocketOptions.query = {};
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Missing required property');
+					done();
+				});
+			});
+		});
+
+		describe('failing with INVALID_HEADERS code and description', function () {
+
+			it('should fail with without port', function (done) {
+				delete validClientSocketOptions.query.port;
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Expected type integer but found type not-a-number');
+					done();
+				});
+			});
+
+			it('should fail without height', function (done) {
+				delete validClientSocketOptions.query.height;
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('#/height: Expected type integer but found type not-a-number');
+					done();
+				});
+			});
+
+			it('should fail without version', function (done) {
+				delete validClientSocketOptions.query.version;
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Missing required property: version');
+					done();
+				});
+			});
+
+			it('should fail without nethash', function (done) {
+				delete validClientSocketOptions.query.nethash;
+				connect();
+				expectDisconnect(this, function (code, description) {
+					expect(code).equal(failureCodes.INVALID_HEADERS);
+					expect(description).contain('Missing required property: nethash');
+					done();
+				});
+			});
+		});
+	});
+
+	describe('with valid headers', function () {
+
+		before(function () {
+			wsServer.start();
+		});
+
+		after(function () {
+			wsServer.stop();
+		});
+
+		beforeEach(function () {
+			/**
+			 * Change of state
+			 * from: not present nonce, not present connectionId, present on master
+			 * to: present nonce, present connectionId, present on master
+			 */
+			validClientSocketOptions.query.nonce = randomString.generate(16);
+			connect();
+		});
+
+		describe('when present on master', function () {
+
+			describe('when nonce is not present', function () {
+
+				beforeEach(function () {
```

Comment:
that's the whole point of these tests - you initialize the connection first with the random nonce A. Because the WS server is running you can be sure that nonce A is a peer already. 
Then in `when nonce is not present` section the case when someone reconnects using different nonce but the same socked id is being checked. Similar thing but with changed socket id is tested in `when nonce is present` section. By changing a query I am enforcing new socket id assignment. Finally, all these steps are repeated by the peer with random nonce which is previously disconnected (state = 1) on master node (it will not appear on the peers list anymore)